### PR TITLE
Raise an exception when getting state at an outlier

### PR DESCRIPTION
**Pull request by richvdh**
Rather than silently returning an invalid result when someone tries to get the state at an outlier event, raise an exception.

~~Based on #12332~~ now rebased

### CODE REVIEW

**Comment by richvdh** [comment]
@DMRobertson I've now rebased this, per your suggestion in #12332.

File: synapse/storage/databases/main/state.py [hunk 1]
```diff
@@ -309,9 +309,13 @@ async def _get_state_group_for_event(self, event_id: str) -> Optional[int]:
         num_args=1,
     )
     async def _get_state_group_for_events(
-        self, event_ids: Iterable[str]
+        self, event_ids: Collection[str]
```
Comment (DMRobertson) [inline synapse/storage/databases/main/state.py:312]: Ahh, because we now iterate over `event_ids` twice.

**Review by DMRobertson — APPROVED** [review]
SGTM.

Do we anticipate this causing any pain for anyone running a homeserver? I'm guessing the only consequence is that something which doesn't work silently will now not work loudly?

File: synapse/storage/databases/main/state.py [hunk 2]
```diff
@@ -307,7 +307,11 @@ async def _get_state_group_for_event(self, event_id: str) -> Optional[int]:
         num_args=1,
     )
     async def _get_state_group_for_events(self, event_ids: Collection[str]) -> JsonDict:
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/storage/databases/main/state.py]: yes, I guess so!

**Comment by richvdh** [comment]
> Do we anticipate this causing any pain for anyone running a homeserver? 

I'm basically hoping I've found and fixed everywhere that we rely on this. But it's possible some instances have slipped through the cracks. We'll have to monitor for exceptions and follow up with any fixes.

File: synapse/storage/databases/main/state.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/storage/databases/main/state.py]: oh wait, this change has already been done on develop.

**Pull request merged**